### PR TITLE
dockerapi: Migrated ImageInspect, ImageLoad, and ImageTag to Docker SDK

### PR DESCRIPTION
**Pull request by kavoor**
### Summary

Migrated ImageInspect, ImageLoad, and ImageTag to Docker SDK
### Implementation details

- Migrated docker.Image to the Docker SDK type ImageInspect
- Migrated ImageInspect, ImageLoad, ContainerKill, and ImageTag to Docker SDK
- Migrated errors to Docker SDK (errNoSuchImage)
- Updated files affected by these migrations to correct types, mocks, and function parameters.
Link to Branch: https://github.com/kavoor/amazon-ecs-agent/tree/imgMoby
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes:  yes

### Description for the changelog

Enhancement - Migrated ImageInspect and ImageLoad to Docker SDK
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/common_integ_test.go [hunk 1]
```diff
@@ -76,12 +76,10 @@ func setup(cfg *config.Config, state dockerstate.TaskEngineState, t *testing.T)
 	if !isDockerRunning() {
 		t.Skip("Docker not running")
 	}
-	ctx, cancel := context.WithCancel(context.TODO())
-	defer cancel()
 
-	clientFactory := clientfactory.NewFactory(ctx, dockerEndpoint)
-	sdkClientFactory := sdkclientfactory.NewFactory(ctx, dockerEndpoint)
-	dockerClient, err := dockerapi.NewDockerGoClient(clientFactory, sdkClientFactory, cfg, ctx)
+	clientFactory := clientfactory.NewFactory(context.Background(), dockerEndpoint)
```
Comment (haikuoliu) [inline agent/engine/common_integ_test.go]: Why is this changed to `context.Background()`?

File: agent/eni/pause/load.go [hunk 2]
```diff
@@ -18,13 +18,13 @@ import (
 
 	"github.com/aws/amazon-ecs-agent/agent/config"
 	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
-	docker "github.com/fsouza/go-dockerclient"
+    "github.com/docker/docker/api/types"
```
Comment (haikuoliu) [inline agent/eni/pause/load.go]: nit: space

File: agent/engine/docker_image_manager_integ_test.go [hunk 3]
```diff
@@ -313,7 +315,7 @@ func TestImageWithSameNameAndDifferentID(t *testing.T) {
 	task2.Containers[0].Image = identicalImageName
 	task3.Containers[0].Image = identicalImageName
 
-	err = renameImage(test3Image1Name, "testimagewithsamenameanddifferentid", "latest", goDockerClient)
+	err = renameImage(context.Background(), test3Image1Name, task1.Containers[0].Image, sdkDockerClient)
```
Comment (haikuoliu) [inline agent/engine/docker_image_manager_integ_test.go]: Can you change this to the following to avoid dangling context? Same below.
```
ctx, cancel := context.WithCancel(context.TODO())
defer cancel()
err = renameImage(ctx, test3Image1Name, task1.Containers[0].Image, sdkDockerClient)
```

File: agent/engine/common_integ_test.go [hunk 4]
```diff
@@ -77,8 +77,8 @@ func setup(cfg *config.Config, state dockerstate.TaskEngineState, t *testing.T)
 		t.Skip("Docker not running")
 	}
 
-	clientFactory := clientfactory.NewFactory(context.Background(), dockerEndpoint)
-	sdkClientFactory := sdkclientfactory.NewFactory(context.Background(), dockerEndpoint)
+	clientFactory := clientfactory.NewFactory(context.TODO(), dockerEndpoint)
```
Comment (haikuoliu) [inline agent/engine/common_integ_test.go]: Please do the same here, `context.TODO()` is essentially same as `context. Background()`: 

```
ctx, cancel := context.WithCancel(context.TODO())
defer cancel()
```

File: agent/engine/docker_image_manager_integ_test.go [hunk 5]
```diff
@@ -530,24 +524,20 @@ func TestImageWithSameIDAndDifferentNames(t *testing.T) {
 
 	// Verify images are removed by docker
 	_, err = taskEngine.(*DockerTaskEngine).client.InspectImage(imageID1)
-	assert.Equal(t, docker.ErrNoSuchImage, err, "Image was not removed successfully")
+	assert.True(t, client.IsErrNotFound(err), "Image was not removed successfully")
 }
 
 // renameImage retag the image and delete the original tag
```
[the remark this answers is not in the document]
Comment (kavoor) [reply agent/engine/docker_image_manager_integ_test.go]: I think the assert statements are doing the same thing. Which comments are out of date? If the images are removed successfully, there will be an IsErrNotFound error, else I print the error message above.

[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_image_manager_integ_test.go]: I mean the comments below, since you updated the function signature:
```
// renameImage retag the image and delete the original tag
```

File: agent/engine/docker_image_manager_integ_test.go [hunk 6]
```diff
@@ -530,24 +524,20 @@ func TestImageWithSameIDAndDifferentNames(t *testing.T) {
...
 
-// renameImage retag the image and delete the original tag
-func renameImage(original, repo, tag string, client *docker.Client) error {
-	err := client.TagImage(original, docker.TagImageOptions{
-		Repo:  repo,
-		Tag:   tag,
-		Force: false,
-	})
+// renameImage retag the image with the target tag and delete the source tag
+func renameImage(ctx context.Context, source string, target string, client *client.Client) error {
```
Comment (sharanyad) [inline agent/engine/docker_image_manager_integ_test.go]: Do we really need to pass in `ctx` to `renameImage`? Can't we just use `context.TODO()` in `client.ImageTag`

File: agent/engine/engine_integ_test.go [hunk 7]
```diff
@@ -91,9 +93,9 @@ func dialWithRetries(proto string, address string, tries int, timeout time.Durat
 }
 
 func removeImage(t *testing.T, img string) {
-	client, err := docker.NewClient(endpoint)
+	client, err := sdk.NewClientWithOpts(sdk.WithHost(endpoint))
 	require.NoError(t, err, "create docker client failed")
-	client.RemoveImage(img)
+	client.ImageRemove(context.Background(), img, types.ImageRemoveOptions{})
```
Comment (sharanyad) [inline agent/engine/engine_integ_test.go]: It is preferable to use `context.TODO` in tests so that it can be canceled.

File: agent/engine/engine_unix_integ_test.go [hunk 8]
```diff
@@ -189,11 +191,9 @@ func TestPortForward(t *testing.T) {
 	// Kill the existing container now to make the test run more quickly.
 	containerMap, _ := taskEngine.(*DockerTaskEngine).state.ContainerMapByArn(testTask.Arn)
 	cid := containerMap[testTask.Containers[0].Name].DockerID
-	client, _ := docker.NewClient(endpoint)
-	err = client.KillContainer(docker.KillContainerOptions{ID: cid})
-	if err != nil {
-		t.Error("Could not kill container", err)
-	}
+	client, _ := sdk.NewClientWithOpts(sdk.WithHost(endpoint))
+	err = client.ContainerKill(context.Background(), cid, "SIGKILL")
```
Comment (sharanyad) [inline agent/engine/engine_unix_integ_test.go]: It is preferable to use context.TODO here and tests below so that it can be canceled.

Comment (kavoor) [reply agent/engine/engine_unix_integ_test.go]: Oh, I had thought I had moved all of the context.Background(), let me fix those.

File: agent/engine/docker_image_manager_integ_test.go [hunk 9]
```diff
@@ -530,24 +524,20 @@ func TestImageWithSameIDAndDifferentNames(t *testing.T) {
...
+func renameImage(source string, target string, client *client.Client) error {
+	err := client.ImageTag(context.TODO(), source, target)
 	if err != nil {
 		return fmt.Errorf("Trying to tag image failed, err: %v", err)
 	}
 
-	// delete the original tag
-	err = client.RemoveImage(original)
+	// delete the source tag
+	_, err = client.ImageRemove(context.TODO(), source, types.ImageRemoveOptions{})
```
Comment (sharanyad) [inline agent/engine/docker_image_manager_integ_test.go]: Instead of directly passing in `context.TODO()`, can you do 

```
ctx, cancel := context.WithCancel(context.TODO())
defer cancel()
``` 

and pass in `ctx`?
